### PR TITLE
#116864429 create channel endpoints

### PR DESCRIPTION
**Pull request by olotintemitope**
#### What does this PR do?

Re-define route and create a controller for channel endpoints on Suyabay API
#### Description of Task to be completed?
- Create Api folder for Suyabay API controllers
- Re-define the channel controller routes
- Create postChannel method in channelController class
- Create getAllChannel method 
- Create getASingleChannel method
- Create editChannel for PUT request
- Create editChannel for PATCH request
#### How should this be manually tested?

Testing the channel endpoints is very easy, all you have to do is to install POSTMAN chrome extension on your browser, then the testing becomes a lot easy. So lets go.

A lot of flexibilities has been implemented here when a user is trying to retrieve the channels. So what next? To get all channel without any query parameters, just type `api/v1/channels` in the url of your POSTMAN and change the request to GET then press send you should see;

```
{
  "data": [
    {
      "user_id": 1,
      "channel_name": "Mrs. Bailee Heathcote PhD",
      "channel_note": "Laboriosam voluptatum nesciunt beatae dolorem neque ut fuga est.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 2,
      "channel_name": "Alexander Koss",
      "channel_note": "Doloribus voluptates et voluptatum sed error et fugiat necessitatibus.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 3,
      "channel_name": "Wilton Schamberger",
      "channel_note": "Et rerum reiciendis impedit at non ipsum.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 4,
      "channel_name": "Mr. Stewart Gibson V",
      "channel_note": "Dolorem harum est saepe ipsam.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 5,
      "channel_name": "Ms. Ruthie Sawayn DVM",
      "channel_note": "Dolorum doloremque numquam voluptatem dolore voluptatem laborum quisquam.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 6,
      "channel_name": "laztopaz",
      "channel_note": "tope0852",
      "subscription_count": 0,
      "date_created": {
        "date": "2016-04-05 12:00:52.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-04-05 12:00:52.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 7,
      "channel_name": "Abuja Kilishis",
      "channel_note": "Abuja Kilishi is one of the best roasted beef meat you will eat and always come back for more",
      "subscription_count": 0,
      "date_created": {
        "date": "2016-04-06 10:21:02.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-04-06 10:24:04.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    }
  ]
}
```
- I have implemented pagination so that developers consuming our API can retrieve results per page. To paginate your result,  just type `api/v1/channels?page=2` in the url of your POSTMAN and change the request to GET then press send you should see;

```
{
  "data": [
    {
      "user_id": 1,
      "channel_name": "Mrs. Bailee Heathcote PhD",
      "channel_note": "Laboriosam voluptatum nesciunt beatae dolorem neque ut fuga est.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 2,
      "channel_name": "Alexander Koss",
      "channel_note": "Doloribus voluptates et voluptatum sed error et fugiat necessitatibus.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 3,
      "channel_name": "Wilton Schamberger",
      "channel_note": "Et rerum reiciendis impedit at non ipsum.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 4,
      "channel_name": "Mr. Stewart Gibson V",
      "channel_note": "Dolorem harum est saepe ipsam.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 5,
      "channel_name": "Ms. Ruthie Sawayn DVM",
      "channel_note": "Dolorum doloremque numquam voluptatem dolore voluptatem laborum quisquam.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 6,
      "channel_name": "laztopaz",
      "channel_note": "tope0852",
      "subscription_count": 0,
      "date_created": {
        "date": "2016-04-05 12:00:52.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-04-05 12:00:52.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 7,
      "channel_name": "Abuja Kilishis",
      "channel_note": "Abuja Kilishi is one of the best roasted beef meat you will eat and always come back for more",
      "subscription_count": 0,
      "date_created": {
        "date": "2016-04-06 10:21:02.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-04-06 10:24:04.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    }
  ]
}
```
- Next thing you can also is to define page limit along side with page query parameter, to do that, just type `api/v1/channels?page=2&results=5` in the url of your POSTMAN and change the request to GET then press send you should see;

```
{
  "data": [
    {
      "user_id": 1,
      "channel_name": "Mrs. Bailee Heathcote PhD",
      "channel_note": "Laboriosam voluptatum nesciunt beatae dolorem neque ut fuga est.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 2,
      "channel_name": "Alexander Koss",
      "channel_note": "Doloribus voluptates et voluptatum sed error et fugiat necessitatibus.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 3,
      "channel_name": "Wilton Schamberger",
      "channel_note": "Et rerum reiciendis impedit at non ipsum.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 4,
      "channel_name": "Mr. Stewart Gibson V",
      "channel_note": "Dolorem harum est saepe ipsam.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 5,
      "channel_name": "Ms. Ruthie Sawayn DVM",
      "channel_note": "Dolorum doloremque numquam voluptatem dolore voluptatem laborum quisquam.",
      "subscription_count": 10,
      "date_created": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-03-10 08:36:38.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 6,
      "channel_name": "laztopaz",
      "channel_note": "tope0852",
      "subscription_count": 0,
      "date_created": {
        "date": "2016-04-05 12:00:52.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-04-05 12:00:52.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    },
    {
      "user_id": 7,
      "channel_name": "Abuja Kilishis",
      "channel_note": "Abuja Kilishi is one of the best roasted beef meat you will eat and always come back for more",
      "subscription_count": 0,
      "date_created": {
        "date": "2016-04-06 10:21:02.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "date_modified": {
        "date": "2016-04-06 10:24:04.000000",
        "timezone_type": 3,
        "timezone": "UTC"
      },
      "created_by": 1
    }
  ]
}
```
-  To get a single Channel then, just type `api/v1/channels/Alexander Koss` in the url of your POSTMAN and change the request to GET then press send you should see;

```
{
  "data": {
    "user_id": 2,
    "channel_name": "Alexander Koss",
    "channel_note": "Doloribus voluptates et voluptatum sed error et fugiat necessitatibus.",
    "subscription_count": 10,
    "date_created": {
      "date": "2016-03-10 08:36:38.000000",
      "timezone_type": 3,
      "timezone": "UTC"
    },
    "date_modified": {
      "date": "2016-03-10 08:36:38.000000",
      "timezone_type": 3,
      "timezone": "UTC"
    },
    "created_by": 1
  }
}
```
- To create a new channel, type `api/v1/channels` then  change your request method to POST. then enter 

```
     channel_name = suyabay
     channel_description = One stop spot for Suya lovers

```

 in the form and press send. Then you should see; `{ 'message' :  'Channel created  successfully' }`
- To edit a channel with a multiple resource, type `api/v1/channels/suyabay` then change your request method to PUT. then enter 

```
     channel_name = suyabay
     channel_description = One stop spot for Suya lovers
```

 in the form and press send. Then you should see; `{ 'message' :  'Channel updated  successfully' }`
- To edit a channel with a single resource, type `api/v1/channels/suyabay` then change your request method to PATCH. then enter 

```
 channel_name = suyabay 

```

 in the form and press send. Then you should see;  `{ 'message' :  'Channel updated  successfully' }`
#### What are the relevant pivotal tracker stories?
#116864429
#### Questions:

[@unicodeveloper](https://github.com/unicodeveloper)

### CODE REVIEW

**Comment by unicodeveloper** [comment]
Please always attach screenshots to the PR @andela-tolotin. It makes the reviewer's work very easy.

**Comment by olotintemitope** [comment]
@unicodeveloper Okay, I will attach screenshots to this PR next time. I'm sorry about this

File: app/Http/Controllers/Api/ChannelController.php [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+     *
+     * @param $request
+     * @param $channel
+     *
+     * @return $response
+     */
+    public function processEditChannel($request, $channel)
+    {
+        if (! is_null($channel)) {
+            DB::table('channels')
```
Comment (andela-oadebayo) [inline app/Http/Controllers/Api/ChannelController.php]: **Code Quality - Nitpick**
@andela-tolotin For the sake of consistency can we stick to using Eloquent and not Query Builder. Please fix this all through your code.

**Comment by unicodeveloper** [comment]
Please fix all the changes we talked about in our team interaction @andela-tolotin

**Comment by olotintemitope** [comment]
@unicodeveloper I am currently working on it. I should be able to finish it up before standup.

**Comment by unicodeveloper** [comment]
Awesome @andela-tolotin

File: app/Http/Controllers/Api/ChannelController.php [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+     * @param $request
+     * @param $response
+     *
+     * @return json $response
+     */
+    public function getAllChannels(Request $request, ChannelTransformer $channelTransformer)
+    {
+        $perPage = $request->query('results') ? : 10;
+
+        $channels = Channel::orderBy('channels.id', 'asc')
```
Comment (OgaBoss) [inline app/Http/Controllers/Api/ChannelController.php]: **Code Quality- Nitpick**
@andela-tolotin Looking at what you want to achieve here, this query is looking too complex. Lets discuss about this.

File: app/Http/Controllers/Api/ChannelController.php [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    /**
+     * This method get a single channel
+     *
+     * @param $channel_name
+     *
+     * @return json $response
+     */
+    public function getAChannel($channel_name, ChannelTransformer $channelTransformer)
+    {
+        $channel = Channel::where('channel_name', '=', strtolower(urldecode($channel_name)))
```
Comment (OgaBoss) [inline app/Http/Controllers/Api/ChannelController.php]: **Code Quality- Nitpick**
@andela-tolotin Looking at what you want to achieve here, this query is looking too complex. Lets discuss about this.

File: app/Http/Controllers/Api/ChannelController.php [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+     *
+     * @return $response
+     */
+    public function deleteASingleChannel(Request $request, $channel_name)
+    {
+        $channel = Channel::where('channel_name', '=', strtolower(urldecode($channel_name)))
+        ->first();
+
+        if (! is_null($channel)) {
+            $returnValue = $this->channelRepository->deleteChannel($channel->id);
```
Comment (OgaBoss) [inline app/Http/Controllers/Api/ChannelController.php:216]: **Code Quality- Nitpick**
@andela-tolotin There is possibility that something goes wrong and the delete will not work, Its better if you do a check here to be sure that the soft delete worked before sending out the message.

**Comment by unicodeveloper** [comment]
Please rebase with the latest on staging @andela-tolotin

**Comment by olotintemitope** [comment]
@unicodeveloper noted, I will do it ASAP

**Comment by olotintemitope** [comment]
@unicodeveloper Done.

**Comment by olotintemitope** [comment]
@unicodeveloper Done.

File: app/Channel.php [hunk 5]
```diff
@@ -38,4 +38,19 @@ public function episode()
     {
         return $this->hasMany('Suyabay\Episode');
     }
+
+    public function scopePendingEpisodes($query, $id)
+    {
+        return $query->where('channels.id', '=', $id)
+                    ->join('episodes', 'channels.id', '=', 'episodes.channel_id')
```
Comment (OgaBoss) [inline app/Channel.php:45]: **Code Quality-Nitpick**
@andela-tolotin Quoting integers in PHP is a ticking time bomb, that might cause future bugs to creep out, please fix this and all other similar occurrences.

File: app/Http/Controllers/Api/ChannelController.php [hunk 6]
```diff
@@ -37,21 +42,43 @@ public function getAllChannels(Request $request, ChannelTransformer $channelTran
...
+
+    }
+
+    public function formatChannel($channels)
+    {
+        foreach ($channels as $key => &$value) {
+            $pending = Channel::pendingEpisodes($value->id)->count();
+            $active  = Channel::activeEpisodes($value->id)->count();
+            $value['user_id'] = UserRepository::findUser($value->user_id)->username;
+            if ($pending != 0 && $active != 0) {
```
Comment (OgaBoss) [inline app/Http/Controllers/Api/ChannelController.php]: **Code Quality-Nitpick**
@andela-tolotin The if statement here will only cater for a scenario where a channel has both active and pending episodes, please refactor to cater for other scenario.

**Pull request merged**